src/ui/public/visualize/components/visualization.tsx

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { get } from 'lodash';
+import React from 'react';
+
+import { PersistedState } from '../../persisted_state';
+import { Vis } from '../../vis';
+import { VisualizationChart } from './visualization_chart';
+import { VisualizationNoResults } from './visualization_noresults';
+
+import './visualization.less';
+
+function shouldShowNoResultsMessage(vis: Vis, visData: any): boolean {
+  const requiresSearch = get(vis, 'type.requiresSearch');
+  const isZeroHits = get(visData, 'hits.total') === 0;
+  const shouldShowMessage = !get(vis, 'params.handleNoResults');
+
+  return Boolean(requiresSearch && isZeroHits && shouldShowMessage);
+}
+
+interface VisualizationProps {
+  listenOnChange: boolean;
+  onInit?: () => void;
+  uiState: PersistedState;
+  vis: Vis;
+  visData: any;
+}
+
+interface VisualizationState {
+  listenOnChange: boolean;
+  showNoResultsMessage: boolean;
+}
+
+export class Visualization extends React.Component<VisualizationProps, VisualizationState> {
+  public static getDerivedStateFromProps(
+    props: VisualizationProps,
+    prevState: VisualizationState
+  ): Partial<VisualizationState> | null {
+    const listenOnChangeChanged = props.listenOnChange !== prevState.listenOnChange;
+    const uiStateChanged = props.uiState && props.uiState !== props.vis.getUiState();
+    if (listenOnChangeChanged || uiStateChanged) {
+      throw new Error('Changing listenOnChange or uiState props is not allowed!');
+    }
+
+    const showNoResultsMessage = shouldShowNoResultsMessage(props.vis, props.visData);
+    if (prevState.showNoResultsMessage !== showNoResultsMessage) {
+      return { showNoResultsMessage };
+    }
+    return null;
+  }
+
+  constructor(props: VisualizationProps) {
+    super(props);
+
+    const { vis, visData, uiState, listenOnChange } = props;
+
+    vis._setUiState(props.uiState);
+    if (listenOnChange) {
+      uiState.on('change', this.onUiStateChanged);
+    }
+
+    this.state = {
+      listenOnChange,
+      showNoResultsMessage: shouldShowNoResultsMessage(vis, visData),
+    };
+  }
+
+  public render() {
+    const { vis, visData, onInit, uiState } = this.props;
+
+    return (
+      <div className="visualization">
+        {this.state.showNoResultsMessage ? (
+          <VisualizationNoResults onInit={onInit} />
+        ) : (
+          <VisualizationChart vis={vis} visData={visData} onInit={onInit} uiState={uiState} />
+        )}
+      </div>
+    );
+  }
+
+  public componentWillUnmount() {
+    this.props.uiState.off('change', this.onUiStateChanged);
```

Comment:
I addressed this issue now in a different way. Since I was able to remove the state completely, using `getDerivedStateFromProps` would anyway be nasty, because as long as that method exists I need to initialize an empty state (even if my typing said there is none), see previous test failure.

So I actually moved the check to `shouldComponentUpdate` and throw the error there. This will also have the sideeffect, that in case we throw there, the properties in `componenWillUnmount` are still the old ones, and we deregister from the correct `uiState`.

Of course the ideal solution would still be allowing that `uiState` switch, but I fear that's a bit too much trouble to change right now, and since I also didn't introduce it with this PR, I would rather postpone this to a later PR/point in time, if it would be okay for you?